### DISCUSSION TITLE
How to make Webhook by status change?

### DISCUSSION THREAD

**Discussion by rynskiy** [directus/directus#15259] [Archive]
Hi,

i newbie by Directus and need some help by this task:

- I have a "Posts" collection with some articles
- one field in this "Posts" collection is "Status" (one to many to collection "Statuses" because standard field Status does not suit me by roles and permissions)
- I have created some webhook by "Webhooks"

I need to start my Webhook every time when Status in collection "Posts" is updated from status "REVIEWING" to status "READY TO PUBLISH". How can I set this? Do I need to create a Flow for this?

Best Regards,
Tony

**Comment by licitdev** [comment]
Yes, flows can trigger a webhook when an item's status is updated to `published`.
Here's an example to get you started.

![image](https://user-images.githubusercontent.com/26413686/186695742-97c6dcec-8d35-434c-8908-c795f23e8f2a.png)

## Steps

1. Create a new flow with an `Event Hook` trigger.
2. Set the scope to be `items.create` and / or `items.update`.
3. Select the collection that you're targeting.
4. Create a new `Condition` operation.
5. Add the following condition rule.
    ```json
    {
        "$trigger": {
            "payload": {
                "status": {
                    "_eq": "published"
                }
            }
        }
    }
    ```
6. Create a new `Webhook / Request URL` for the `Resolve` route.

**Reply by rynskiy** [reply]
Thanks @licitdev, but how can I set a condition: if status updated **from status _EDITING_ to _READY TO PUBLISH_**? What's important for me to start webhook ONLY if status changed from EDITING to READY TO PUBLISH and not just by updating from anything to READY TO PUBLISH.

**Reply by licitdev** [reply]
You will have to change the event hook type to `filter`, as that triggers before the update is performed.

## Changes
Add a `Read Data` after the same condition check as before to read the current status.
Add a new condition check that checks if the status is `draft`.
```json
{
    "$last": {
        "status": {
            "_eq": "draft"
        }
    }
}
```
Finally proceed with the webhook.

![image](https://user-images.githubusercontent.com/26413686/186713965-4cb8abfe-48d3-491b-87c6-1fa40a98daee.png)

Note: For the `Response Body` option in the trigger, you can leave it unselected or to be set to `{{$trigger.payload}}` as you're not changing the payload.

Kindly refer to https://docs.directus.io/configuration/flows.html for more details on Flows, triggers and operations.

Just in case you weren't aware of, you can customise the status field interface to allow options other than `published`, `draft` and `archived`.

**Comment by rynskiy** [comment]
Thanks a lot @licitdev! Unfortunately, by the first example too, 

The Flow started correct (I can see this in Log), but I don't receive a webhook by Integromat :( What have I done wrong?
Here are my screens;
![Unbenannt2](https://user-images.githubusercontent.com/54798086/186915159-f7c9ec9b-1989-4d5e-8ea5-6920ea3252bb.PNG)
![Unbenannt3](https://user-images.githubusercontent.com/54798086/186915177-d4498171-13ce-43df-9419-1a8dbb1f7ae6.PNG)
![Unbenannt1](https://user-images.githubusercontent.com/54798086/186915763-2cbe0b65-1c2f-4abe-abbd-025289542277.PNG)

**Reply by licitdev** [reply]
@rynskiy As shown in the log, the status in `$trigger.payload` is `9` and not `READY TO PUBLISH`.
Thus the webhook did not run. If it ran, the webhook operation log will be displayed.
The above was an example and it will have to be adapted to your data structure accordingly. 👍

**Reply by rynskiy** [reply]
@licitdev thanks, with status 9 it seems in Log now what webhook will be started, but with Error... What is ""Request failed with status code 400", "name": "AxiosError"
![Unbenannt4](https://user-images.githubusercontent.com/54798086/186926498-71c9360c-5adb-41f1-acad-bc610cd01dd3.PNG)

**Reply by licitdev** [reply]
The error 400 is coming from Integromat. It seems that the webhook to Integromat wasn't configured correctly. Perhaps you might be missing some headers or the form body has to be in a certain format. Kindly refer to their docs.

**Reply by rynskiy** [reply]
Thanks @licitdev! It seems to work correctly now with Request Body in form:
[
		{"id": "id"}
]

Can you please explain to me how can I send now with webhook for example a "ID" from triggered post? Unfortunately, I don't know yet a syntax of Directus. 
Must it be:? 
[
		{"id": "{{ $trigger.payload.id }}"}
]

Actually, all I need is just to send with webhook an ID of article with changed Status...

**Reply by rynskiy** [reply]
Do you have some docs with Directus syntax for Request Body? I need to understand the logic of variables like {{ $trigger.collection.item}} :)

**Reply by licitdev** [reply]
You may use `$last` to accesss the data from the previous operation, or the `key` of the operation.
Please refer to the docs, the available keys can be found in  https://docs.directus.io/configuration/flows.html#the-flow-object 

For example, you may have an operation that has the key `read_article`.

![image](https://user-images.githubusercontent.com/26413686/186941262-e80c0c99-5e26-464a-b69c-b2c58b6fd2e0.png)

You may refer to the data in another operation referenced with the `read_article` key.
You can use the dot notation and array notation to access the data within.
Eg: `{{read_article.id}}` or `{{read_article.authors[0]}}`

You may also check out the `Run Script` operation released in `9.16.0` that allows you to run  JavaScript code. 
Docs for this operation is coming soon~ 🙂

**Reply by rynskiy** [reply]
hmm... Ok. But I can't still understand, how can I send with my webhook the ID of my Article with changed to "Ready to Publish"=9 status... Can you please explain how must it be in Request Body? Because [	{ "id": "{{ $last.id }}" }] doesn't work....

**Reply by rynskiy** [reply]
@licitdev I have just 3 modules in my Flow (please see screens). Can you please explain what must be in Request Body to bekame in webhook the ID of article which have changed a status to Ready to Publish?
![flow1](https://user-images.githubusercontent.com/54798086/186952377-be65178e-b677-43bf-a8ca-a9391cf8538a.PNG)
![flow2](https://user-images.githubusercontent.com/54798086/186952373-649ce28f-54b6-455e-bd30-04582af9e48f.PNG)
![flow3](https://user-images.githubusercontent.com/54798086/186952366-722c38ed-79ac-49fe-a3ca-0f3e3f15ba46.PNG)

**Reply by licitdev** [reply]
You will have to refer to your logs to see what fields are available for use. 

Here's an example of a Trigger log.
![image](https://user-images.githubusercontent.com/26413686/186969133-ac301234-ac60-461e-bd87-814ed4649a87.png)
To retrieve the ID of the item: `{{$trigger.keys[0]}}`
To retrieve the updated description: `{{$trigger.payload.description}}`
